File: Scala/src/test/scala/ComparisonTest.scala
import org.scalatest._
import org.scalatest.selenium.Firefox
import org.scalatest.concurrent.Eventually._
import org.scalatest.time.{Millis, Seconds, Span}

class ComparisonTest extends FlatSpec with Firefox with Matchers with GivenWhenThen {

  implicit val patienceConfig =
    PatienceConfig(timeout = scaled(Span(10, Seconds)), interval = scaled(Span(200, Millis)))
  val pattern = "\\d{3},\\d{2}"

  "The lowest price of Venue 7 tablet on Allegro" should "be higher than 350,00 PLN" in {
    When("Going to Allegro Page")
    go to "http://allegro.pl"

    Then("title should start with Allegro.pl")
    pageTitle should startWith("Allegro.pl")

    When("We display list of Venue 7 offers in Tablets category")
    textField("main-search-text").value = "venue 7\n"
    click on xpath("//span[text() = 'Komputery']")
    click on xpath("//span[@class='name' and contains(.,'Tablety')]")

    And("sort results by lowest price")
    click on cssSelector("div.toggle span.label")
    eventually {
      click on xpath("//dt[text()='cena']/following-sibling::dd[1]/descendant::a[contains(.,'od najniższej')]")
    }
    eventually {
      xpath("//span[@class='label' and contains(.,'od najniższej')]").element.isDisplayed should be(true)
    }

    Then("the lowest price of device should be higher than 350,00 PLN")
    val priceStrings = findFirstThreeDigitPrice
    priceStrings should not be empty
    val amount = priceStrings.get.replace(",", ".").toDouble
    amount should be > 350.00

    quit()
  }

  private def findFirstThreeDigitPrice: Option[String] = {
    findAll(xpath("//*[@id='featured-offers']/descendant::article/descendant::span[@class='buy-now dist']")).map {
      case node => pattern.r.findFirstIn(node.text)
    }.filter(_.isDefined).toList.head
  }
}

File: Java/src/test/java/ComparisonTest.java
import org.junit.Test;
import org.openqa.selenium.By;
import org.openqa.selenium.WebElement;
import org.openqa.selenium.firefox.FirefoxDriver;
import org.openqa.selenium.support.ui.ExpectedConditions;
import org.openqa.selenium.support.ui.WebDriverWait;

import java.util.List;
import java.util.Optional;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

import static org.assertj.core.api.Assertions.assertThat;

public class ComparisonTest extends FirefoxDriver {

    // given
    private int timeoutInSeconds = 10;
    private int intervalInMilliseconds = 200;
    private Pattern amountRegex = Pattern.compile("(\\d{3},\\d{2})");

    @Test
    public void checkTheLowestPriceOfVenue7OnAllegro() {
        //when
        get("http://allegro.pl");

        //then
        getTitle().startsWith("Allegro.pl");

        //when
        findElement(By.id("main-search-text")).sendKeys("venue 7\n");
        findElement(By.xpath("//span[text() = 'Komputery']")).click();
        findElement(By.xpath("//span[@class='name' and contains(.,'Tablety')]")).click();

        //and
        findElement(By.cssSelector("div.toggle span.label")).click();
        waitForClickableElement(By
                .xpath("//dt[text()='cena']/following-sibling::dd[1]/descendant::a[contains(.,'od najniższej')]"))
                .click();
        waitForElementToDisplayText(By.cssSelector("div.toggle span.label"), "od najniższej");

        //then
        Optional<Double> amount = findFirstThreeDigitPrice();
        assertThat(amount.isPresent()).isTrue();
        assertThat(amount.get()).isGreaterThan(350.0);

        quit();
    }

    private WebElement waitForClickableElement(By by) {
        return new WebDriverWait(this, timeoutInSeconds, intervalInMilliseconds).until(ExpectedConditions
                .elementToBeClickable(by));
    }

    private void waitForElementToDisplayText(By by, String expectedText) {
        new WebDriverWait(this, timeoutInSeconds, intervalInMilliseconds).until(ExpectedConditions
                .textToBePresentInElementLocated(by, expectedText));
    }

    private Optional<Double> findFirstThreeDigitPrice() {
        List<WebElement> prices = findElements(By
                .xpath("//*[@id='featured-offers']/descendant::article/descendant::span[@class='buy-now dist']"));
        return prices.stream()
                .filter(price -> amountRegex.matcher(price.getText()).find()).map(price -> {
                    Matcher matcher = amountRegex.matcher(price.getText());
                    matcher.find();
                    return Double.valueOf(matcher.group(1).replace(",", "."));
                }).findFirst();
    }
}
